### PR TITLE
Support cluster mode when using Redis as a store

### LINKED ISSUE #869: Redis Store Connection should support Cluster and Connection Manager
Currently the Redis store supports Connection Manager, but it should also support redis cluster.

### PR DESCRIPTION
**Pull request by caass**
# Description

This PR brings support for using Redis in cluster mode as a store.

Fixes #869

## Type of change

Please delete options that aren't relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
  not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Notably, no new tests have been added. Since we currently use a mock to test our redis support, we don't have infrastructure in place to test against a live redis cluster.

I'd like to add some tests that spin up a redis cluster and assert that everything works ok in cluster mode, but it's a non-trivial amount of work. I'd like for them to run using the rust test harness, so likely that'd mean grouping them under a cargo feature (like, "slow-tests") and using testcontainers to create a redis cluster.

## Checklist

- [x] Updated documentation if needed
- [ ] Tests added/amended
- [x] `bazel test //...`  passes locally
- [x] PR is contained in a single commit, using `git amend` see some [docs](https://www.atlassian.com/git/tutorials/rewriting-history)

- - -
This change is [Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/998)

### CODE REVIEW

**Review by caass — COMMENTED** [review]
I didn't add tests yet because it felt like a lot of story points to build out infrastructure for running tests against actual redis instances. If necessary, I can block this PR on getting that built out (and ensuring tests pass with a single instance), and then unblock this once that infrastructure lands in main. I'm kind of scared because the reasoning about which commands are allowed in cluster mode vs. which aren't and whether or not you can use a pipeline or transaction aren't enforced at compile time, so I have a lot less confidence in "if it compiles, it works."

> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/nativelink/nativelink/998#-:-O-HJ34A4V1dWn2-NpB4:brxuoxr)* status: 0 of 1 LGTMs obtained, and pending CI: Analyze (javascript-typescript), Analyze (python), Bazel Dev / ubuntu-22.04, Cargo Dev / macos-13, Cargo Dev / ubuntu-22.04, Installation / macos-13, Installation / macos-14, Installation / ubuntu-22.04, Local / ubuntu-22.04, Publish image, Publish nativelink-worker-lre-cc, asan / ubuntu-22.04, docker-compose-compiles-nativelink (20.04), docker-compose-compiles-nativelink (22.04), integration-tests (20.04), integration-tests (22.04), macos-13, pre-commit-checks, ubuntu-20.04 / stable, ubuntu-22.04, ubuntu-22.04 / stable, vale, windows-2022 / stable, and 3 discussions need to be resolved

___
*[`nativelink-store/src/redis_store.rs` line 68 at r1](https://reviewable.io/reviews/nativelink/nativelink/998#-O-HJWkSBXI0T2lvRC-B:-O-HJWkSBXI0T2lvRC-C:b-npd1rw) ([raw file](https://github.com/nativelink/nativelink/blob/89c97db77f85fa90f19e82e6cd123f3bb0c7a492/nativelink-store/src/redis_store.rs#L68)):*
> ```XML
> // I (@caass) tried a little to make this work but lost a couple days on it and decided to timebox.
> #[derive(Clone)]
> pub enum NativelinkRedisConnection {
> ```

Nit: This doesn't need to be `pub`, it's a holdover from when I was trying to test cluster support before I concluded it was a bigger job than I was comfortable doing in the same PR as this change.
___
*[`nativelink-store/src/redis_store.rs` line 68 at r1](https://reviewable.io/reviews/nativelink/nativelink/998#-O-HJgA-62K012g8pNnx:-O-HJgA-62K012g8pNny:b-resha3) ([raw file](https://github.com/nativelink/nativelink/blob/89c97db77f85fa90f19e82e6cd123f3bb0c7a492/nativelink-store/src/redis_store.rs#L68)):*
> ```XML
> // I (@caass) tried a little to make this work but lost a couple days on it and decided to timebox.
> #[derive(Clone)]
> pub enum NativelinkRedisConnection {
> ```

I spent a long time trying to avoid building this wrapper enum; I wanted to just use `RedisStore<dyn ConnectionLike>` instead but it kept getting confusing trying to pass around this trait object since we swap it into cells, move it to and from the stack, and all sorts of things during initialization. I was digging into `ArcCell`, it was a rabbit hole.
___
*[`nativelink-store/src/redis_store.rs` line 106 at r1](https://reviewable.io/reviews/nativelink/nativelink/998#-O-HKCG6-q-Ey2tINZXj:-O-HKCG6-q-Ey2tINZXk:b-pa9h0e) ([raw file](https://github.com/nativelink/nativelink/blob/89c97db77f85fa90f19e82e6cd123f3bb0c7a492/nativelink-store/src/redis_store.rs#L106)):*
> ```XML
>         }
>     }
> }
> ```

FYI there's crates like [delegate](https://docs.rs/delegate/latest/delegate/) that remove this boilerplate, but I don't think this is so horrible yet that it warrants bringing in a new dependency.



_Code quote:_
```XML
impl ConnectionLike for NativelinkRedisConnection {
    fn req_packed_command<'a>(
        &'a mut self,
        cmd: &'a redis::Cmd,
    ) -> redis::RedisFuture<'a, redis::Value> {
        match self {
            NativelinkRedisConnection::Single(inner) => inner.req_packed_command(cmd),
            NativelinkRedisConnection::Cluster(inner) => inner.req_packed_command(cmd),
        }
    }

    fn req_packed_commands<'a>(
        &'a mut self,
        cmd: &'a redis::Pipeline,
        offset: usize,
        count: usize,
    ) -> redis::RedisFuture<'a, Vec<redis::Value>> {
        match self {
            NativelinkRedisConnection::Single(inner) => {
                inner.req_packed_commands(cmd, offset, count)
            }
            NativelinkRedisConnection::Cluster(inner) => {
                inner.req_packed_commands(cmd, offset, count)
            }
        }
    }

    fn get_db(&self) -> i64 {
        match self {
            NativelinkRedisConnection::Single(inner) => inner.get_db(),
            NativelinkRedisConnection::Cluster(inner) => inner.get_db(),
        }
    }
}
```
___
*[`nativelink-store/src/redis_store.rs` line 110 at r1](https://reviewable.io/reviews/nativelink/nativelink/998#-O-HKR-71OPmo9FoGB1m:-O-HKR-71OPmo9FoGB1n:bxaqqen) ([raw file](https://github.com/nativelink/nativelink/blob/89c97db77f85fa90f19e82e6cd123f3bb0c7a492/nativelink-store/src/redis_store.rs#L110)):*
> ```XML
> impl NativelinkRedisConnection {
>     /// Create a new connection to a single redis instance
>     pub async fn single<T: redis::IntoConnectionInfo>(params: T) -> Result<Self, Error> {
> ```

Nit: Similar comment about the `pub` here.
___
*[`nativelink-store/src/redis_store.rs` line 121 at r1](https://reviewable.io/reviews/nativelink/nativelink/998#-O-HKTq43hn7Pkz7AKVE:-O-HKTq43hn7Pkz7AKVF:b-966ih1) ([raw file](https://github.com/nativelink/nativelink/blob/89c97db77f85fa90f19e82e6cd123f3bb0c7a492/nativelink-store/src/redis_store.rs#L121)):*
> ```XML
> 
>     /// Create a connection to a redis cluster
>     pub async fn cluster<T: redis::IntoConnectionInfo>(
> ```

Nit: and here.
___
*[`nativelink-store/src/redis_store.rs` line 248 at r1](https://reviewable.io/reviews/nativelink/nativelink/998#-O-HKetGCoP43vhO2suz:-O-HKetGCoP43vhO2sv-:bs2k0k2) ([raw file](https://github.com/nativelink/nativelink/blob/89c97db77f85fa90f19e82e6cd123f3bb0c7a492/nativelink-store/src/redis_store.rs#L248)):*
> ```XML
> 
>                     async {
>                         let mut conn = self.get_conn().await.err_tip(|| {
> ```

The call to `get_conn` is moved inside the async block here, rather than reused. But since each call eventually becomes a `Arc::get`, I don't think it's a huge deal.

**Review by caass — COMMENTED** [review]
+@allada

> *[Reviewable](https://reviewable.io/reviews/nativelink/nativelink/998#-:-O-HLMiv79uH43l-w9vC:b-aeorft)* status: 0 of 1 LGTMs obtained, and pending CI: Analyze (javascript-typescript), Analyze (python), Bazel Dev / ubuntu-22.04, Cargo Dev / macos-13, Cargo Dev / ubuntu-22.04, Installation / macos-13, Installation / macos-14, Installation / ubuntu-22.04, Local / ubuntu-22.04, Publish image, Publish nativelink-worker-lre-cc, asan / ubuntu-22.04, docker-compose-compiles-nativelink (20.04), docker-compose-compiles-nativelink (22.04), integration-tests (20.04), integration-tests (22.04), macos-13, pre-commit-checks, ubuntu-20.04 / stable, ubuntu-22.04, ubuntu-22.04 / stable, vale, windows-2022 / stable, and 3 discussions need to be resolved (waiting on @allada)

**Review by allada — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tracemachina/nativelink/998)* status: 0 of 1 LGTMs obtained, and pending CI: Cargo Dev / macos-13, Publish image, Publish nativelink-worker-lre-cc, and 7 discussions need to be resolved

___
*[`-- commits` line 7 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-Iaxid639JCoAwVTwV):*
nit: Remove everything here and below, it's not relevant to this change. This would normally just be a comment in a ticket or in the review.
___
*[`nativelink-store/src/redis_store.rs` line 68 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-HJWkSBXI0T2lvRC-B:-O-IbP7C3ud6KRRTNEKE:b-aprfhf) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L68)):*
<details><summary><i>Previously, caass (Cass Fridkin) wrote…</i></summary><blockquote>

Nit: This doesn't need to be `pub`, it's a holdover from when I was trying to test cluster support before I concluded it was a bigger job than I was comfortable doing in the same PR as this change.
</blockquote></details>

Yeah, lets remove pub if possible.
___
*[`nativelink-store/src/redis_store.rs` line 68 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-HJgA-62K012g8pNnx:-O-IhqbFDSZyzXkTe4-E:b6y1d9c) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L68)):*
<details><summary><i>Previously, caass (Cass Fridkin) wrote…</i></summary><blockquote>

I spent a long time trying to avoid building this wrapper enum; I wanted to just use `RedisStore<dyn ConnectionLike>` instead but it kept getting confusing trying to pass around this trait object since we swap it into cells, move it to and from the stack, and all sorts of things during initialization. I was digging into `ArcCell`, it was a rabbit hole.
</blockquote></details>

Yeah, i took a small stab at it, it's a pain to work around. Wrapping seems fine.
___
*[`nativelink-store/src/redis_store.rs` line 77 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-IbzTs15SHU_qOQ8OZ:-O-IbzTs15SHU_qOQ8O_:bvlu8r9) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L77)):*
> ```XML
>         &'a mut self,
>         cmd: &'a redis::Cmd,
>     ) -> redis::RedisFuture<'a, redis::Value> {
> ```

nit: `'_` didnt work?
___
*[`nativelink-store/src/redis_store.rs` line 106 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-HKCG6-q-Ey2tINZXj:-O-IiE2ABrT-egn-kZkL:b-j40jzx) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L106)):*
<details><summary><i>Previously, caass (Cass Fridkin) wrote…</i></summary><blockquote>

FYI there's crates like [delegate](https://docs.rs/delegate/latest/delegate/) that remove this boilerplate, but I don't think this is so horrible yet that it warrants bringing in a new dependency.
</blockquote></details>

This is a neat crate, we might use it in some other areas.
___
*[`nativelink-store/src/redis_store.rs` line 110 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-HKR-71OPmo9FoGB1m:-O-IiLl606kw2AteK9e-:b80c393) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L110)):*
<details><summary><i>Previously, caass (Cass Fridkin) wrote…</i></summary><blockquote>

Nit: Similar comment about the `pub` here.
</blockquote></details>

nit: remove `pub` if we don't need it.
___
*[`nativelink-store/src/redis_store.rs` line 121 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-HKTq43hn7Pkz7AKVE:-O-IiOit3nEGOVyPAgI8:b-rqxvo7) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L121)):*
<details><summary><i>Previously, caass (Cass Fridkin) wrote…</i></summary><blockquote>

Nit: and here.
</blockquote></details>

ditto.
___
*[`nativelink-store/src/redis_store.rs` line 138 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-IlH6N-H0rZrPD30ZW:-O-IlH6N-H0rZrPD30ZX:b4c0qfb) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L138)):*
> ```XML
>         let conn_fut = if config.addresses.len() == 1 {
>             let addr = config.addresses[0].clone();
>             NativelinkRedisConnection::single(addr).boxed().shared()
> ```

I don't believe this is what single/cluster means. There's an actual different mode redis will be in and I believe the clients must know if the server is a cluster server or single server.

We probably need to introduce a config that the user needs to specify if redis is a cluster or non-cluster, but please read/check this.
___
*[`nativelink-store/src/redis_store.rs` line 248 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-HKetGCoP43vhO2suz:-O-Ij9ieAE8b8qTN__j7:bdt8eem) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L248)):*
<details><summary><i>Previously, caass (Cass Fridkin) wrote…</i></summary><blockquote>

The call to `get_conn` is moved inside the async block here, rather than reused. But since each call eventually becomes a `Arc::get`, I don't think it's a huge deal.
</blockquote></details>

I would suggest  doing it more similar to what it used to do. This reduces the need for a bunch of needless branches, since we will have a connection, we can just clone it before passing it into the future.
___
*[`nativelink-store/src/redis_store.rs` line 258 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-Ik-7tEff2_SWP-JSt:-O-Ik-7tEff2_SWP-JSu:b-19j7hi) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L258)):*
> ```XML
>                     }
>                 })
>                 .collect::<FuturesOrdered<_>>();
> ```

Instead lets use [.buffered()](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.buffered)

And limit it to something like 50. This will help reduce an edge case where we have an insane number of futures outstanding all fighting each other.

**Review by adam-singer — CHANGES_REQUESTED** [review]
> Reviewed 4 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/tracemachina/nativelink/998)* status: 0 of 1 LGTMs obtained, and pending CI: Cargo Dev / macos-13, Publish image, Publish nativelink-worker-lre-cc, and 10 discussions need to be resolved (waiting on @allada)

___
*[`nativelink-store/src/redis_store.rs` line 61 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-KLAJxEcZwNwWJzQfN:-O-KLAJxEcZwNwWJzQfO:b75cmt5) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L61)):*
> ```XML
> /// and [cluster](https://redis.io/docs/latest/operate/oss_and_stack/reference/cluster-spec/) modes.)
> //
> // Theoretically, we shouldn't need this as [`RedisStore`] uses an [`ArcCell`] internally to hold its connection.
> ```

Are we trying to convey this is something we should change or a future `TODO`? If so lets make it a `TODO(name):`.

I think we are getting to a point where `TODO` should probably be ticketed and tracked every time they happen. We are not fully there yet.
___
*[`nativelink-store/src/redis_store.rs` line 66 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-KL0ZTEH5lOVdQleI9:-O-KL0ZTEH5lOVdQleIA:bja7ik0) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L66)):*
> ```XML
> // always inside a [`Result`] enum, whose arguments must be sized.
> //
> // I (@caass) tried a little to make this work but lost a couple days on it and decided to timebox.
> ```

This comment doesn't seem relevant on a technical level, lets remove.
___
*[`nativelink-store/src/redis_store.rs` line 197 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-KLyQcBlJqljZsfo1T:-O-KLyQcBlJqljZsfo1U:b-1ac8r7) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L197)):*
> ```XML
> 
>     /// Encode a [`StoreKey`] so it can be sent to Redis.
>     // This doesn't really need to return an `impl`, but I want to avoid relying specifically on
> ```

Lets drop the use of `I` in the code comments and use a `TODO(name)` if it intended to be refactored at a later time. Generally doc comments shouldn't use `I` as many people are working on the code base.

**Review by caass — COMMENTED** [review]
Ok, made some pretty sweeping changes here. Whoops...

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tracemachina/nativelink/998#-:-O-bNSz42mI_cNvUH_JD:bx5fdc5)* status: 0 of 1 LGTMs obtained, and pending CI: Bazel Dev / ubuntu-22.04, Cargo Dev / macos-13, Cargo Dev / ubuntu-22.04, Installation / macos-13, Installation / macos-14, Installation / ubuntu-22.04, Publish image, Publish nativelink-worker-lre-cc, asan / ubuntu-22.04, docker-compose-compiles-nativelink (20.04), docker-compose-compiles-nativelink (22.04), macos-13, ubuntu-22.04, windows-2022 / stable, and 6 discussions need to be resolved (waiting on @allada)

___
*[`nativelink-store/src/redis_store.rs` line 61 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-KLAJxEcZwNwWJzQfN:-O-bEAWg8XFzsHTAMNTa:b-iy3c1u) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L61)):*
<details><summary><i>Previously, adam-singer (Adam Singer) wrote…</i></summary><blockquote>

Are we trying to convey this is something we should change or a future `TODO`? If so lets make it a `TODO(name):`.

I think we are getting to a point where `TODO` should probably be ticketed and tracked every time they happen. We are not fully there yet.
</blockquote></details>

Marking this as resolved b/c the code changed pretty significantly.
___
*[`nativelink-store/src/redis_store.rs` line 66 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-KL0ZTEH5lOVdQleI9:-O-bE5N_59-rka4puebJ:b-iy3c1u) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L66)):*
<details><summary><i>Previously, adam-singer (Adam Singer) wrote…</i></summary><blockquote>

This comment doesn't seem relevant on a technical level, lets remove.
</blockquote></details>

Marking this as resolved b/c the code changed pretty significantly.
___
*[`nativelink-store/src/redis_store.rs` line 77 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-IbzTs15SHU_qOQ8OZ:-O-bEEse2VIFZ4dqqT2c:b-83q4go) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L77)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

nit: `'_` didnt work?
</blockquote></details>

No :(
___
*[`nativelink-store/src/redis_store.rs` line 138 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-IlH6N-H0rZrPD30ZW:-O-bEObS2NY47uFhRBM0:bhmjrls) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L138)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

I don't believe this is what single/cluster means. There's an actual different mode redis will be in and I believe the clients must know if the server is a cluster server or single server.

We probably need to introduce a config that the user needs to specify if redis is a cluster or non-cluster, but please read/check this.
</blockquote></details>

Yeah, you're right about this.
___
*[`nativelink-store/src/redis_store.rs` line 197 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-KLyQcBlJqljZsfo1T:-O-bET5Y18PK08tdq-8B:bim2pwq) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L197)):*
<details><summary><i>Previously, adam-singer (Adam Singer) wrote…</i></summary><blockquote>

Lets drop the use of `I` in the code comments and use a `TODO(name)` if it intended to be refactored at a later time. Generally doc comments shouldn't use `I` as many people are working on the code base.
</blockquote></details>

Ok, will do.
___
*[`nativelink-store/src/redis_store.rs` line 258 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-Ik-7tEff2_SWP-JSt:-O-bMMsa0zWW4ifVxRcO:berssk9) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L258)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

Instead lets use [.buffered()](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.buffered)

And limit it to something like 50. This will help reduce an edge case where we have an insane number of futures outstanding all fighting each other.
</blockquote></details>

I'm running into a problem here trying to use [`futures::stream::iter`](https://docs.rs/futures/latest/futures/stream/fn.iter.html) where we actually define the `StoreDriver` trait such that the relationship between the lifetime of the borrowed slice of `StoreKey`s and the data inside those `StoreKey`s is undefined -- that is to say, according to the rust compiler, the data inside the `StoreKey`s could change before the reference is made invalid. A more correct definition would look like this:

```rust

#[async_trait]
pub trait StoreDriver: Sync + Send + Unpin + HealthStatusIndicator + 'static {
    /// See: [`StoreLike::has_with_results`] for details.
    async fn has_with_results<'a, 'b: 'a>(
        self: Pin<&Self>,
        digests: &'a [StoreKey<'b>],
        results: &mut [Option<usize>],
    ) -> Result<(), Error>;
}
```

At least, I think that's the problem...the actual error message I'm getting is:
```text
error: implementation of `std::ops::FnOnce` is not general enough
   --> nativelink-store/src/redis_store.rs:345:28
    |
345 |       ) -> Result<(), Error> {
    |  ____________________________^
346 | |         if keys.len() == 1 && is_zero_digest(keys[0].borrow()) {
347 | |             results[0] = Some(0);
348 | |             return Ok(());
...   |
391 | |         Ok(())
392 | |     }
    | |_____^ implementation of `std::ops::FnOnce` is not general enough
    |
    = note: closure with signature `fn((usize, &nativelink_util::store_trait::StoreKey<'0>)) -> {async block@nativelink-store/src/redis_store.rs:359:13: 369:14}` must implement `std::ops::FnOnce<((usize, &nativelink_util::store_trait::StoreKey<'1>),)>`, for any two lifetimes `'0` and `'1`...
    = note: ...but it actually implements `std::ops::FnOnce<((usize, &nativelink_util::store_trait::StoreKey<'_>),)>`

```

which is kind of...nonsensical.

**Review by adam-singer — APPROVED** [review]
:lgtm:

> Reviewed 2 of 3 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tracemachina/nativelink/998#-:-O-guoio4cnfVENsViXi:bnfp4nl)* status: 0 of 1 LGTMs obtained, and 4 discussions need to be resolved (waiting on @allada)

___
*[`nativelink-config/src/stores.rs` line 898 at r3](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-gdC7k4pB2R55ZtyA-:-O-gdC7k4pB2R55ZtyA0:bkpmji0) ([raw file](https://github.com/tracemachina/nativelink/blob/3dc31092fe6e3e67dd6d589513eb69aec1f05c6e/nativelink-config/src/stores.rs#L898)):*
> ```XML
>     /// Default: false,
>     #[serde(default)]
>     pub cluster: bool,
> ```

Would an `enum` type be a better option for different modes or connection types?

**Review by caass — COMMENTED** [review]
> Reviewed 3 of 3 files at r4, all commit messages.
> Dismissed @allada from a discussion.
> *[Reviewable](https://reviewable.io/reviews/tracemachina/nativelink/998)* status: 0 of 1 LGTMs obtained, and pending CI: Local / ubuntu-22.04, Vercel, pre-commit-checks, vale, and 1 discussions need to be resolved (waiting on @allada)

___
*[`nativelink-config/src/stores.rs` line 898 at r3](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-gdC7k4pB2R55ZtyA-:-O-gvH_w57sidQIO76iO:b-sp1flf) ([raw file](https://github.com/tracemachina/nativelink/blob/3dc31092fe6e3e67dd6d589513eb69aec1f05c6e/nativelink-config/src/stores.rs#L898)):*
<details><summary><i>Previously, adam-singer (Adam Singer) wrote…</i></summary><blockquote>

Would an `enum` type be a better option for different modes or connection types?
</blockquote></details>

Potentially, the only other option seems to be [Sentinel](https://docs.rs/redis/latest/redis/sentinel/index.html) So we can make that happen. Done.

**Review by adam-singer — APPROVED** [review]
> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/tracemachina/nativelink/998)* status: 0 of 1 LGTMs obtained, and pending CI: Bazel Dev / ubuntu-22.04, Cargo Dev / macos-13, Cargo Dev / ubuntu-22.04, Installation / macos-13, Installation / macos-14, Installation / ubuntu-22.04, Local / ubuntu-22.04, Publish image, Publish nativelink-worker-lre-cc, docker-compose-compiles-nativelink (20.04), docker-compose-compiles-nativelink (22.04), integration-tests (20.04), integration-tests (22.04), macos-13, pre-commit-checks, windows-2022 / stable, and 1 discussions need to be resolved (waiting on @allada)

**Review by zbirenbaum — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/tracemachina/nativelink/998#-:-O-hI2rB93UVfkhQvXP9:bnfp4nl)* status: 0 of 1 LGTMs obtained, and pending CI: Cargo Dev / macos-13, Installation / macos-13, Installation / macos-14, Local / ubuntu-22.04, integration-tests (20.04), integration-tests (22.04), macos-13, and 1 discussions need to be resolved (waiting on @allada)

**Review by caass — COMMENTED** [review]
> Dismissed @allada from a discussion.
> *[Reviewable](https://reviewable.io/reviews/tracemachina/nativelink/998)* status: 0 of 1 LGTMs obtained (waiting on @allada)

___
*[`nativelink-store/src/redis_store.rs` line 258 at r1](https://reviewable.io/reviews/tracemachina/nativelink/998#-O-Ik-7tEff2_SWP-JSt:-O-kBsUQ28A2SRisB9f3:bqw1j2f) ([raw file](https://github.com/tracemachina/nativelink/blob/9e16976caa940d6795e865203a2ec0c02d0a3e93/nativelink-store/src/redis_store.rs#L258)):*
<details><summary><i>Previously, caass (Cass Fridkin) wrote…</i></summary><blockquote>

I'm running into a problem here trying to use [`futures::stream::iter`](https://docs.rs/futures/latest/futures/stream/fn.iter.html) where we actually define the `StoreDriver` trait such that the relationship between the lifetime of the borrowed slice of `StoreKey`s and the data inside those `StoreKey`s is undefined -- that is to say, according to the rust compiler, the data inside the `StoreKey`s could change before the reference is made invalid. A more correct definition would look like this:

```rust

#[async_trait]
pub trait StoreDriver: Sync + Send + Unpin + HealthStatusIndicator + 'static {
    /// See: [`StoreLike::has_with_results`] for details.
    async fn has_with_results<'a, 'b: 'a>(
        self: Pin<&Self>,
        digests: &'a [StoreKey<'b>],
        results: &mut [Option<usize>],
    ) -> Result<(), Error>;
}
```

At least, I think that's the problem...the actual error message I'm getting is:
```text
error: implementation of `std::ops::FnOnce` is not general enough
   --> nativelink-store/src/redis_store.rs:345:28
    |
345 |       ) -> Result<(), Error> {
    |  ____________________________^
346 | |         if keys.len() == 1 && is_zero_digest(keys[0].borrow()) {
347 | |             results[0] = Some(0);
348 | |             return Ok(());
...   |
391 | |         Ok(())
392 | |     }
    | |_____^ implementation of `std::ops::FnOnce` is not general enough
    |
    = note: closure with signature `fn((usize, &nativelink_util::store_trait::StoreKey<'0>)) -> {async block@nativelink-store/src/redis_store.rs:359:13: 369:14}` must implement `std::ops::FnOnce<((usize, &nativelink_util::store_trait::StoreKey<'1>),)>`, for any two lifetimes `'0` and `'1`...
    = note: ...but it actually implements `std::ops::FnOnce<((usize, &nativelink_util::store_trait::StoreKey<'_>),)>`

```

which is kind of...nonsensical.
</blockquote></details>

I'm going to address this in a follow-up PR once the trait is modified.

**Pull request merged**
